application-load-balancer/lambda_function/secret_rotator.py

```diff
@@ -0,0 +1,625 @@
+                'FieldToMatch': {'SingleHeader': {'Name': self.header_name}},
+                'PositionalConstraint': 'EXACTLY',
+                'SearchString': search_string,
+                'TextTransformations': [{'Type': 'NONE', 'Priority': 0}]
+            }
+        }
+
+    def update_waf_acl(self, new_secret: str, prev_secret: str) -> None:
+        client = boto3.client('wafv2')
+        waf_acl = self.get_waf_acl()
+        lock_token = waf_acl['LockToken']
+        metric_name = f"{self.application}-{self.environment}-XOriginVerify"
+        rule = {
+            'Name': f"{self.application}{self.environment}XOriginVerify",
+            'Priority': self.waf_rule_priority,
+            'Action': {'Allow': {}},
+            'VisibilityConfig': {
+                'SampledRequestsEnabled': True,
+                'CloudWatchMetricsEnabled': True,
+                'MetricName': metric_name
+            },
+            'Statement': {
+                'OrStatement': {
+                    'Statements': [
+                        self._create_byte_match_statement(new_secret),
+                        self._create_byte_match_statement(prev_secret)
+                    ]
+                }
+            }
+        }
+
+        new_rules = [rule] + [r for r in waf_acl['WebACL']['Rules'] if r['Priority'] != self.waf_rule_priority]
+        logger.info("Updating WAF WebACL with new rules.")
+
+        response = client.update_web_acl(
+            Name=self.waf_acl_name,
+            Scope='REGIONAL',
+            Id=self.waf_acl_id,
+            LockToken=lock_token,
+            DefaultAction={'Block': {}},
+            Description='CloudFront Origin Verify',
+            VisibilityConfig={
+                'SampledRequestsEnabled': True,
+                'CloudWatchMetricsEnabled': True,
+                'MetricName': metric_name
+            },
+            Rules=new_rules
+        )
+
+        if response['ResponseMetadata']['HTTPStatusCode'] == 200:
+            logger.info("WAF WebACL rules updated")
+
+    def get_cf_distro(self, distro_id: str) -> Dict:
+        """
+        Fetches the CloudFront distribution details.
+        """
+        client = self.get_cloudfront_client()
+        return client.get_distribution(Id=distro_id)
+
+
+    def get_cf_distro_config(self, distro_id: str) -> Dict:
+        """
+        Fetches the configuration of a CloudFront distribution.
+        """
+        client = self.get_cloudfront_client()
+        return client.get_distribution_config(Id=distro_id)
+
+
+    def update_cf_distro(self, distro_id: str, header_value: str) -> Dict:
+        """
+        Updates the custom headers for a CloudFront distribution.
+
+        Args:
+            distro_id (str): The ID of the CloudFront distribution.
+            header_value (str): The header value to set for the custom header.
+        """
+        client = self.get_cloudfront_client()
+
+        if not self.is_distribution_deployed(distro_id):
```

Comment:
If you are only ever updating *deployed* distributions and erroring out if it's not deployed, then maybe you could determine if the distribution is deployed in the first instance, in the `get_distro_list ` function? (rename to `get_deployed_distributions`?)  This would remove a bit of the later logic, checking whether it's deployed or not and make it easier to see what's going on.  I'm not sure at this point if you are sometimes doing stuff with not deployed distributions though!

Reply:
This can be picked up in a later PR.  We want to maintain the structure for now so we dont break our tests.